### PR TITLE
mw_basic_auth: avoid missing header errors when using body basic auth…

### LINKED ISSUE #2193: bug: body basic auth credentials
**Do you want to request a *feature* or report a *bug*?**

bug

**What is the current behavior?**

when extracting credentials from body is enabled, we still look in headers first - as such we always get info level log `Attempted access with malformed header, no auth header found.`.

**What is the expected behavior?**

We should not see this log message.

**Which versions of Tyk affected by this issue? Did this work in previous versions of Tyk?**

master

### PR DESCRIPTION
**Pull request by matiasinsaurralde**
Fix #2193

### CODE REVIEW

File: mw_basic_auth.go [hunk 1]
```diff
@@ -77,6 +77,11 @@ func (k *BasicAuthKeyIsValid) requestForBasicAuth(w http.ResponseWriter, msg str
 func (k *BasicAuthKeyIsValid) basicAuthHeaderCredentials(w http.ResponseWriter, r *http.Request) (username, password string, err error, code int) {
 	token := r.Header.Get("Authorization")
 	logger := k.Logger().WithField("key", obfuscateKey(token))
+	if token == "" && k.Spec.BasicAuth.ExtractFromBody {
```
Comment (dencoded) [inline mw_basic_auth.go]: this edge case is a good candidate to cover by test

Comment (matiasinsaurralde) [reply mw_basic_auth.go]: Seems that we already have tests with empty headers + body extraction enabled in `TestBasicAuthFromBody`.

Comment (dencoded) [reply mw_basic_auth.go]: why that test wasn't failing without this fix then?

**Comment by matiasinsaurralde** [comment]
@dencoded I've replaced this with a simpler solution, now we should discuss if we want to log auth errors on `basicAuthBodyCredentials` in the same way we do for `basicAuthHeaderCredentials`, think this should be a separate issue/PR.

File: mw_basic_auth.go [hunk 2]
```diff
@@ -79,7 +79,9 @@ func (k *BasicAuthKeyIsValid) basicAuthHeaderCredentials(w http.ResponseWriter,
 	logger := k.Logger().WithField("key", obfuscateKey(token))
 	if token == "" {
 		// No header value, fail
-		logger.Info("Attempted access with malformed header, no auth header found.")
+		if !k.Spec.BasicAuth.ExtractFromBody {
+			logger.Info("Attempted access with malformed header, no auth header found.")
```
Comment (dencoded) [inline mw_basic_auth.go]: I think this message should be ERROR or WARN level (not info).
also, it seems to be extra have it here with reference to out auth extraction from body.
how about moving this log message to `ProcessRequest` and put this warning in log in appropriate situation ( `if k.Spec.BasicAuth.ExtractFromBody` is already there).

**Comment by dencoded** [comment]
@matiasinsaurralde I left comments. also, CI is failing again.

File: mw_basic_auth.go (hunk 2 above)
Comment (matiasinsaurralde) [reply mw_basic_auth.go]: @dencoded Updated, can you review?

Comment (matiasinsaurralde) [reply mw_basic_auth.go]: Think we don't need to check for `token` with the new change, this scenario will occur when `basicAuthHeaderCredentials` fails.

**Comment by matiasinsaurralde** [comment]
The CI error isn't related to this change, looks like a linking issue:
```
/home/travis/.gimme/versions/go1.11.8.linux.amd64/pkg/tool/linux_amd64/link: running gcc failed: exit status 1
/usr/bin/ld: /tmp/go-link-995235108/000016.o: unrecognized relocation (0x2a) in section `.text'
/usr/bin/ld: final link failed: Bad value
collect2: error: ld returned 1 exit status
```

**Comment by dencoded** [comment]
@matiasinsaurralde yes, coprocess build is failing

**Comment by dencoded** [comment]
@matiasinsaurralde yes, coprocess build is failing

**Comment by dencoded** [comment]
@matiasinsaurralde yes, coprocess build is failing

**Comment by dencoded** [comment]
was closed by exident

**Comment by buger** [comment]
From code point of view looks good. Coprocess builds seems to be unrelated, and looks like CI issue.
